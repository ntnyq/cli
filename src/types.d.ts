export interface ISelectItem {
  label: string
  value: string
  url?: string
  action?: () => void
}
