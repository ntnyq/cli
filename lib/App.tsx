'use strict'

import React from 'react'
import { Box, Text } from 'ink'
import open from 'open'
import Gradient from 'ink-gradient'
import BigText from 'ink-big-text'
import Link from 'ink-link'
import SelectInput from 'ink-select-input'
import type { ISelectItem } from './types'

const handleSelect = (item: ISelectItem) => {
  if (item.url) return open(item.url)
  item.action?.()
}

const createItems = (items: ISelectItem[]) =>
  items.map(item => ({ ...item, key: item.url || item.label }))

const items: ISelectItem[] = createItems([
  {
    label: `Website`,
    url: `https://ntnyq.com`,
  },
  {
    label: `Twitter`,
    url: `https://twitter.com/ntnyq`,
  },
  {
    label: `Github`,
    url: `https://github.com/ntnyq`,
  },
  {
    label: `Blog`,
    url: `https://ntnyq.github.io`,
  },
  {
    label: `Segment Fault`,
    url: `https://segmentfault.com/u/ntnyq`,
  },
  {
    label: `Quit`,
    action() {
      process.exit()
    },
  },
])

export interface Props {
  name: string
}

export const App = ({ name }: Props) => (
  <Box flexDirection='column'>
    <Gradient name='rainbow'>
      <BigText text={name} />
    </Gradient>
    <Box marginBottom={1}>
      <Link url={`https://${name}.com`}>
        Hello, this is <Text color='cyan'>{name}</Text>, an FE developer.
      </Link>
    </Box>
    <SelectInput items={items as any} onSelect={handleSelect} />
    <Box
      justifyContent='center'
      width='40%'
      borderStyle='classic'
      borderColor='#0ff'
      marginTop={1}
      paddingY={1}
      paddingX={1}
    >
      <Text>
        Copyright &copy; <Text color='green'>{name}</Text>
      </Text>
    </Box>
  </Box>
)
