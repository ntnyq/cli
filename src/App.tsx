'use strict'

import process from 'node:process'
import { Box, Text } from 'ink'
import BigText from 'ink-big-text'
import Gradient from 'ink-gradient'
import Link from 'ink-link'
import SelectInput from 'ink-select-input'
import { UncontrolledTextInput } from 'ink-text-input'
import open from 'open'
import React, { useState } from 'react'
import type { ISelectItem } from './types'

const createItems = (items: Omit<ISelectItem, 'value'>[]) =>
  items.map(item => ({
    ...item,
    value: item.label,
  }))

export interface Props {
  name: string
}

export const App: React.FC<Props> = ({ name }) => {
  const [isComment, setIsComment] = useState(false)

  const items: ISelectItem[] = createItems([
    {
      label: 'Website',
      url: 'https://ntnyq.com',
    },
    {
      label: 'GitHub',
      url: 'https://github.com/ntnyq',
    },
    {
      label: 'Twitter',
      url: 'https://twitter.com/ntnyq',
    },
    {
      label: 'Bilibili',
      url: 'https://space.bilibili.com/40199161',
    },
    {
      label: 'Bluesky',
      url: 'https://bsky.app/profile/ntnyq.com',
    },
    {
      label: 'Juejin',
      url: 'https://juejin.cn/user/1591748566717591',
    },
    {
      label: 'Segment Fault',
      url: 'https://segmentfault.com/u/ntnyq',
    },
    {
      label: 'Leave a message',
      action() {
        setIsComment(true)
      },
    },
    {
      label: 'Quit',
      action() {
        process.exit()
      },
    },
  ])

  const handleSelect = (item: ISelectItem) => {
    if (item.url) {
      return open(item.url)
    }
    item.action?.()
  }
  const handleSubmit = (message: string) => {
    if (message.length) {
      open(
        `https://github.com/ntnyq/ntnyq/issues/new?title=${encodeURIComponent(message)}`,
      )
    }
    return setIsComment(false)
  }

  return isComment ? (
    <Box
      flexDirection='column'
      marginY={1}
    >
      <Box marginBottom={1}>
        <Text>Enter to submit a message or go back:</Text>
      </Box>
      <UncontrolledTextInput
        placeholder='A message to me'
        onSubmit={handleSubmit}
      />
    </Box>
  ) : (
    <Box flexDirection='column'>
      <Gradient name='rainbow'>
        <BigText text={name} />
      </Gradient>
      <Box marginBottom={1}>
        <Link
          url={`https://${name}.com`}
          fallback
        >
          Hello, this is <Text color='cyan'>{name}</Text>, an FE developer.
        </Link>
      </Box>
      <SelectInput
        items={items}
        onSelect={handleSelect}
      />
      <Box
        justifyContent='center'
        width='40%'
        borderStyle='classic'
        borderColor='#0ff'
        marginTop={1}
        paddingY={1}
        paddingX={1}
      >
        <Text>
          Copyright &copy; <Text color='green'>{name}</Text>
        </Text>
      </Box>
    </Box>
  )
}
