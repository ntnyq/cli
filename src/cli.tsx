#!/usr/bin/env node
'use strict'

import { render } from 'ink'
import React from 'react'
import { App } from './App'

render(<App name='ntnyq' />)
