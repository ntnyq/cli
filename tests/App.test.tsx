import process from 'node:process'
import { render } from 'ink-testing-library'
import React from 'react'
import stripAnsi from 'strip-ansi'
import { expect, it } from 'vitest'
import { App } from '../src/App'

it('render App', () => {
  // @ts-expect-error expect a string
  process.env.FORCE_COLOR = 1

  const { lastFrame } = render(<App name='ntnyq' />)

  const stripedFrame = stripAnsi(lastFrame())

  expect(
    ['ntnyq', 'Website', 'Twitter', 'GitHub', 'Bilibili', 'Juejin', 'Segment Fault', 'Quit'].every(
      word => stripedFrame.includes(word),
    ),
  ).toBeTruthy()

  delete process.env.FORCE_COLOR
})

it('snapshot', () => {
  const { lastFrame } = render(<App name='ntnyq' />)
  const stripedFrame = stripAnsi(lastFrame())

  // no `❯` for windows
  expect(stripedFrame.replace('>', '❯')).toMatchSnapshot()
})
