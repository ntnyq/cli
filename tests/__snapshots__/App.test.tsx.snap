// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot 1`] = `
"

 ███╗   ██╗ ████████╗ ███╗   ██╗ ██╗   ██╗  ██████╗
 ████╗  ██║ ╚══██╔══╝ ████╗  ██║ ╚██╗ ██╔╝ ██╔═══██╗
 ██╔██╗ ██║    ██║    ██╔██╗ ██║  ╚████╔╝  ██║   ██║
 ██║╚██╗██║    ██║    ██║╚██╗██║   ╚██╔╝   ██║▄▄ ██║
 ██║ ╚████║    ██║    ██║ ╚████║    ██║    ╚██████╔╝
 ╚═╝  ╚═══╝    ╚═╝    ╚═╝  ╚═══╝    ╚═╝     ╚══▀▀═╝


Hello, this is ntnyq, an FE developer. (​https://ntnyq.com​)

❯ Website
  Twitter
  GitHub
  Bilibili
  Juejin
  Segment Fault
  Leave a message
  Quit

+--------------------------------------+
|                                      |
|          Copyright © ntnyq           |
|                                      |
+--------------------------------------+"
`;
